website_sale_hide_empty_category/models/product_public_category.py

```diff
@@ -7,15 +7,14 @@
 class ProductPublicCategory(models.Model):
     _inherit = "product.public.category"
 
-    has_product_recursive = fields.Boolean(
+    count_product_recursive = fields.Integer(
         string="This category or one of its children has products",
-        compute="_compute_has_product_recursive",
+        compute="_compute_count_product_recursive",
     )
 
-    @api.depends("product_tmpl_ids", "child_id.has_product_recursive")
-    def _compute_has_product_recursive(self):
+    @api.depends("product_tmpl_ids", "child_id.count_product_recursive")
+    def _compute_count_product_recursive(self):
         for category in self:
-            category.has_product_recursive = bool(
-                category.product_tmpl_ids
-                or any(child.has_product_recursive for child in category.child_id)
-            )
+            category.count_product_recursive = len(
+                category.product_tmpl_ids.filtered("is_published")
```

Comment:
Why do you need to change all this stuff when this is the only change needed to make this improvement work?

Reply:
Not sure I understand what you mean. Which stuff you have in mind? Names of the property and compute function? Well to match what they do. Or you have something else in mind?

Reply:
@chienandalu I think I understand now after playing with _website_sale_product_brand_ and trying to make it un/publishable. That `.filtered("is_published")` part is not really necessary as it is added through _ir.rule_ as need be (public user only seeing published products).

The question remains whether you object renaming that property and method. Should I leave the original names that suggest they are of boolean type? Please comment and I would change as necessary.

Reply:
I think it would be less disruptive to just add the product counter as a complementary field. Keep in mind that custom extensions could be broken removing a field just in the middle of a version. The same goes with the views. Custom themes could rely on the current xml architecture, and no ecommerce holder will be happy when a 500 raises after updating the module.

You can add some migration notes anyway to remove the boolean in the next version.

Reply:
But in such case wouldn't a separate module be more appropriate option? But then I am not sure about performance? Would separate (from boolean evaluation) counting affect performance or would ORM handle this without any penalty?

Or how about  just changing the return type and leave the function names intact so that it wouldn't break anything?

Or maybe (that looks the best for me) something like this - a mix of both:
- change the return type of the compute function to integer
- add complementary product counter that just takes the result of the compute function

What would you suggest? Thanx a lot.

Reply:
I was going to propose that, but I think the two fields can coexist in this module. You can compute the two fields altogether and let the boolean be set by the counter (`bool(count_result)`). The compute doesn't return anything anyway, so you can add that behavior.

So the changes would be to keep everything as it is now adding the counter field to the compute.

Then you can just add the custom view extension, as it's not going to harm anything pre-existing

Maybe for v14 (you can put it on the roadmap) the module can be split on two:

- The one that renders the count
- The one that hides the category depending on that count.

Reply:
Thanx a lot for input. I am not sure about v14 - I'd rather suggest renaming the module. Would seem to me that splitting it would be creating very small modules... But that can be decided later on I guess. Going to adjust the module now.